packages/story-editor/src/app/taxonomy/taxonomyProvider.tsx

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2021 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * External dependencies
+ */
+import type {
+  Story,
+  Term,
+  Taxonomy,
+  TermSlug,
+} from '@googleforcreators/elements';
+import {
+  useCallback,
+  useEffect,
+  useMemo,
+  useState,
+} from '@googleforcreators/react';
+import type { PropsWithChildren } from 'react';
+
+/**
+ * Internal dependencies
+ */
+import type {
+  createTermProps,
+  setTermsProps,
+  TaxonomyRestError,
+  addTermToSelectionProps,
+  addSearchResultsToCacheProps,
+  removeTermsProps,
+  UpdateStoryProps,
+} from '../../types';
+import cleanForSlug from '../../utils/cleanForSlug';
+import { useAPI } from '../api';
+import { useStory } from '../story';
+import Context from './context';
+
+function TaxonomyProvider(props: PropsWithChildren<unknown>) {
+  // Should grab categories on mount
+  const [shouldRefetchCategories, setShouldRefetchCategories] =
+    useState<boolean>(true);
+  const { updateStory, terms } = useStory((state) => ({
+    updateStory: state.actions.updateStory,
+    terms: state.state.story?.terms || [],
+  }));
+  const [hasTaxonomies, setHasTaxonomies] = useState<boolean>(false);
+  const [taxonomies, setTaxonomies] = useState<Taxonomy[]>([]);
+  const [termCache, setTermCache] = useState<Term[]>(terms);
+
+  const {
+    actions: { getTaxonomyTerm, createTaxonomyTerm, getTaxonomies },
+  } = useAPI();
+
+  // Get all registered `web-story` taxonomies.
+  useEffect(() => {
+    if (hasTaxonomies || !getTaxonomies) {
+      return;
+    }
+
+    void (async () => {
+      try {
+        const result = await getTaxonomies();
+        setTaxonomies(result);
+      } catch (e) {
+        if (e instanceof Error) {
+          // eslint-disable-next-line no-console -- Log error
+          console.error(e.message);
+        }
+      } finally {
+        setHasTaxonomies(true);
+      }
+    })();
+  }, [hasTaxonomies, getTaxonomies]);
+
+  const setTerms = useCallback(
+    ({ newTerms }: setTermsProps) => {
```

Comment:
No need for passing an object, can just do
```suggestion
  const setTerms = useCallback(
    (newTerms: Term[]) => {
```